### ISSUE TITLE
Upgrade onXXX callback signatures for non-nullable types

### ISSUE THREAD

**Issue by falsandtru** [microsoft/TypeScript#9715]
All of the onXXX callback signatures should be nullable such as `((ev: Event) => any) | null` for assignabillity of null (e.g. `el.onclick = null`), but current signatures are non-nullable.

**TypeScript Version:**  master

**Code**

``` diff
-    onclick: (this: this, ev: MouseEvent) => any;
+    onclick: ((this: this, ev: MouseEvent) => any) | null;
```

related: #9548

cc @zhengbli

**Comment by mhegazy** [comment]
I could not find reference in the spec for this, can you point us to the spec?

**Comment by falsandtru** [comment]
Probably:

> An event handler has a value, which is either null, or is a callback object, or is an internal raw uncompiled handler. The EventHandler callback function type describes how this is exposed to scripts. Initially, an event handler's value must be set to null.

https://html.spec.whatwg.org/multipage/webappapis.html#event-handler-attributes

**Comment by falsandtru** [comment]
@zhengbli @mhegazy @DanielRosenwasser Can you accept this PR for the quoted spec?

**Comment by mhegazy** [comment]
so why not all event handlers? what is special about the `onClick` one?

**Comment by falsandtru** [comment]
All diffs are too large to write in here. This bug is needed to fix officially like #9548. I think `onclick` is a most popular callback that has `on` prefix. So that diff is just an example.

**Comment by mhegazy** [comment]
i am not sure i see the value of a one-off change. we should do them all for consistency and correctness.

**Comment by mhegazy** [comment]
@zhengbli what is the best way to make this change on the lib generator side?

**Comment by falsandtru** [comment]
Of course, fixes of #9548 (`addEventListener` methods) and this (`on-` callbacks) should be synced because these are pairs.

**Comment by Andarist** [comment]
This one was fixed a long time ago. All `.onXXX` properties are nullable in the DOM lib. cc @jakebailey
